core/src/main/java/com/linecorp/armeria/internal/common/DelegatingConnectionEventListener.java

```diff
@@ -0,0 +1,184 @@
+import com.linecorp.armeria.internal.common.util.ChannelUtil;
+
+import io.netty.channel.Channel;
+import io.netty.util.AttributeKey;
+import io.netty.util.AttributeMap;
+
+public final class DelegatingConnectionEventListener {
+
+    private static final AttributeKey<DelegatingConnectionEventListener> CONNECTION_EVENT_LISTENER_KEY =
+            AttributeKey.valueOf(DelegatingConnectionEventListener.class, "CONNECTION_EVENT_LISTENER_KEY");
+
+    @Nullable
+    public static DelegatingConnectionEventListener getOrNull(Channel channel) {
+        return channel.attr(CONNECTION_EVENT_LISTENER_KEY).get();
+    }
+
+    public static DelegatingConnectionEventListener get(Channel channel) {
+        final DelegatingConnectionEventListener listener = getOrNull(channel);
+        if (listener != null) {
+            return listener;
+        }
+        throw new IllegalStateException("DelegatingConnectionEventListener not found");
+    }
+
+    private static final Logger logger = LoggerFactory.getLogger(DelegatingConnectionEventListener.class);
+
+    private final ConnectionEventListener delegate;
+    private final Channel channel;
+    private final InetSocketAddress remoteAddress;
+    private final SessionProtocol desiredProtocol;
+
+    @Nullable
+    private InetSocketAddress localAddress;
+    @Nullable
+    private SessionProtocol actualProtocol;
+
+    public DelegatingConnectionEventListener(ConnectionEventListener delegate, Channel channel,
+                                             SessionProtocol desiredProtocol, InetSocketAddress remoteAddress) {
+        this.delegate = delegate;
+        this.channel = channel;
+        this.remoteAddress = remoteAddress;
+        this.desiredProtocol = desiredProtocol;
+        channel.attr(CONNECTION_EVENT_LISTENER_KEY).set(this);
+    }
+
+    private InetSocketAddress remoteAddress() {
+        return remoteAddress;
+    }
+
+    /**
+     * Lazily acquires and returns the local address of the {@link #channel} since it is valid only after the
+     * connection is established.
+     */
+    private InetSocketAddress localAddress() {
+        if (localAddress == null) {
+            localAddress = ChannelUtil.localAddress(channel);
+        }
+        checkState(localAddress != null, "localAddress not available");
+        return localAddress;
+    }
+
+    private SessionProtocol desiredProtocol() {
+        return desiredProtocol;
+    }
+
+    private AttributeMap attrs() {
+        return channel;
+    }
+
+    /**
+     * Lazily acquires and returns the actual protocol of the {@link #channel} since it is valid only after the
+     * connection is established.
+     */
+    private SessionProtocol actualProtocol() {
+        if (actualProtocol == null) {
+            actualProtocol = HttpSession.get(channel).protocol();
+        }
+        checkState(actualProtocol != null, "actualProtocol not available");
+        return actualProtocol;
+    }
+
+    public void connectionPending() {
```

Comment:
`connectionPending` sounds a little odd since `ClientConnectionTimings#pendingAcquisition` seems to be measuring a different timing although it has a similar name.

Would `sessionPending` make more sense? @line/dx

Reply:
I prefer a method name starting with `connection` to be consistent with other methods.
How about `connectionNegotiating()`?